File: web/src/lib/pk-import.ts
/**
 * PluralKit import — preview calls.
 *
 * The actual import is enqueued via lib/imports.ts (the async job
 * runner). What's left here is the synchronous preview step used by
 * the import flow's review screen.
 */
import { apiFetch } from "./api-client";

export interface PKPreviewMember {
  id: string;
  name: string;
}

export interface PKPreviewSummary {
  system_name: string | null;
  member_count: number;
  members: PKPreviewMember[];
  group_count: number;
  switch_count: number;
  earliest_switch: string | null;
  latest_switch: string | null;
  limit_warnings: string[];
}

export async function previewImportFromFile(file: File): Promise<PKPreviewSummary> {
  const form = new FormData();
  form.append("file", file);
  return apiFetch<PKPreviewSummary>("/v1/import/pluralkit/preview", {
    method: "POST",
    headers: {},
    body: form,
  });
}

export async function previewImportFromApi(token: string): Promise<PKPreviewSummary> {
  return apiFetch<PKPreviewSummary>("/v1/import/pluralkit-api/preview", {
    method: "POST",
    body: JSON.stringify({ token }),
  });
}

File: sheaf/schemas/pk_import.py
"""Pydantic models for PluralKit data import.

Both ingestion paths (file upload and live API pull) produce the same
canonical PK-shaped dict, so a single set of schemas describes the
preview/options/result regardless of source.
"""

from datetime import datetime

from pydantic import BaseModel, Field


class PKImportOptions(BaseModel):
    """What to import from a PluralKit export."""

    system_profile: bool = True
    member_ids: list[str] | None = Field(
        default=None,
        description=(
            "PK member HIDs to import. None = all. Used to let the user "
            "deselect specific members on the preview screen."
        ),
    )
    groups: bool = True
    front_history: bool = False  # Default off — switch logs can run thousands of entries.


class PKApiCredentials(BaseModel):
    """Token-based credentials for the live PluralKit API.

    The token is request-scoped: it lives only for the duration of the
    request handler and is never logged or persisted. The frontend
    submits it via HTTPS POST body.
    """

    token: str = Field(min_length=1, max_length=128)


class PKApiPreviewRequest(PKApiCredentials):
    pass


class PKApiImportRequest(PKApiCredentials):
    options: PKImportOptions = Field(default_factory=PKImportOptions)


class PKPreviewMember(BaseModel):
    id: str  # PK HID
    name: str


class PKPreviewSummary(BaseModel):
    system_name: str | None = None
    member_count: int = 0
    members: list[PKPreviewMember] = []
    group_count: int = 0
    switch_count: int = 0
    earliest_switch: datetime | None = None
    latest_switch: datetime | None = None


class PKImportResult(BaseModel):
    members_imported: int = 0
    groups_imported: int = 0
    fronts_imported: int = 0
    warnings: list[str] = []
